File: src/main/java/controllers/cashbox/CashboxViewController.java
package controllers.cashbox;

import javafx.application.Platform;
import javafx.collections.FXCollections;
import javafx.collections.ObservableList;
import javafx.event.ActionEvent;
import javafx.fxml.FXML;
import javafx.fxml.FXMLLoader;
import javafx.scene.Node;
import javafx.scene.Parent;
import javafx.scene.Scene;
import javafx.scene.control.*;
import javafx.scene.control.cell.PropertyValueFactory;
import javafx.stage.Modality;
import javafx.stage.Stage;
import models.Operation;
import models.Position;
import models.Product;
import models.ProductsData;
import network.Api;
import network.RetrofitClient;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import start.Main;

import java.io.IOException;
import java.util.List;

/**
 * Created by andrey on 24.07.17.
 */
public class CashboxViewController {
    private Main mainApp;

    Api api = RetrofitClient.getApiService();

    private ObservableList<Operation> operationOL = FXCollections.observableArrayList();
    private List<Operation> operationList;
    private List<Product> productList;


    @FXML
    private Label networkStatus;
    @FXML
    private TableView<Operation> operationTable;
    @FXML
    private TableColumn<Operation, String> typeColumn;
    @FXML
    private TableColumn<Operation, String> dateColumn;
    @FXML
    private TableColumn<Operation, String> userColumn;
    @FXML
    private Label totalLable;





    @FXML
    private Label userLable;
    @FXML
    private Label typeLable;
    @FXML
    private Label dateLable;


    @FXML
    private TableView<Position> positionTable;
    @FXML
    private TableColumn<Position, String> nameColumn;
    @FXML
    private TableColumn<Position, Float> priceColumn;
    @FXML
    private TableColumn<Position, String> unitColumn;
    @FXML
    private TableColumn<Position, String> countColumn;
    @FXML
    private TableColumn<Position, String> discountColumn;
    @FXML
    private TableColumn<Position, String> sumColumn;


    @FXML
    private void initialize() {

        nameColumn.setCellValueFactory(new PropertyValueFactory<Position, String>("productName"));
        priceColumn.setCellValueFactory(new PropertyValueFactory<Position, Float>("priceFormat"));
        countColumn.setCellValueFactory(new PropertyValueFactory<Position, String>("countString"));
        discountColumn.setCellValueFactory(new PropertyValueFactory<Position, String>("discountString"));
        sumColumn.setCellValueFactory(new PropertyValueFactory<Position, String>("sumFormat"));
        unitColumn.setCellValueFactory(new PropertyValueFactory<Position, String>("unit"));
        //---------------------------------------------------------------------------------------
        typeColumn.setCellValueFactory(new PropertyValueFactory<Operation, String>("typeString"));
        dateColumn.setCellValueFactory(new PropertyValueFactory<Operation, String>("dateFormat"));
        userColumn.setCellValueFactory(new PropertyValueFactory<Operation, String>("user"));
        operationTable.getSelectionModel().selectedItemProperty().addListener(
                (observable, oldValue, newValue) -> showOperation(newValue));
        loadOperationData();
        loadProducts();
        //---------------------------------------------------------------------------------------
    }

    private void loadOperationData() {
        Call<List<Operation>> call = api.operations();
        call.enqueue(new Callback<List<Operation>>() {
            @Override
            public void onResponse(Call<List<Operation>> call, Response<List<Operation>> response) {
                if (response.code() == 200) {
                    operationList = response.body();
                    operationOL = FXCollections.observableArrayList(response.body());
                    Platform.runLater(new Runnable() {
                        @Override
                        public void run() {
                            operationTable.setItems(operationOL);
                            networkStatus.setText("ONLINE");
                        }
                    });

                } else {
                    networkStatus.setText("OFFLINE");
                }
            }

            @Override
            public void onFailure(Call<List<Operation>> call, Throwable t) {
                networkStatus.setText("OFFLINE");
            }
        });
    }

    public void newOperation(ActionEvent actionEvent){
        try{
            Stage stage = new Stage();
            FXMLLoader loader = new FXMLLoader();
            loader.setLocation(getClass().getResource("../../view/cashbox/NewCashboxView.fxml"));
            Parent categoryAddFXML = loader.load();
            stage.setTitle("Кассовая операция");
            stage.setMinHeight(600);
            stage.setMinWidth(800);
            stage.setResizable(false);
            stage.setScene(new Scene(categoryAddFXML));
            stage.initModality(Modality.WINDOW_MODAL);
            stage.initOwner(((Node)actionEvent.getSource()).getScene().getWindow());
            NewCashboxController controller = loader.getController();
            stage.showAndWait();

            if (controller.isOkClicked()) {
                loadOperationData();
            }

        } catch (IOException e){
            e.printStackTrace();
        }
    }
    private void showOperation(Operation operation){
        userLable.setText(operation.getUser());
        typeLable.setText(operation.getTypeString());
        dateLable.setText(operation.getDateFormat());
        Double sum = 0.0;
        List<Position> positions = operation.getPositions();
        for (int i = 0; i < positions.size(); i++){
            Integer id = positions.get(i).getProduct();
            if (getProductById(id)!=null){
                Product product = getProductById(id);
                positions.get(i).setProductName(product.getName());
                setSum(positions.get(i));
                positions.get(i).setUnit(product.getUnit());
                sum = sum + positions.get(i).getSum();
            }
        }
        ObservableList<Position> positionObservableList = FXCollections.observableArrayList(operation.getPositions());
        positionTable.setItems(positionObservableList);
        totalLable.setText(String.format("%.2f", sum));
    }

    private void loadProducts(){
        Call<ProductsData> call = api.productsData();
        call.enqueue(new Callback<ProductsData>() {
            @Override
            public void onResponse(Call<ProductsData> call, Response<ProductsData> response) {
                if (response.code() == 200) {
                    productList = response.body().getProducts();
                    Platform.runLater(new Runnable() {
                        @Override
                        public void run() {
                            networkStatus.setText("ONLINE");
                        }
                    });

                } else {
                    networkStatus.setText("OFFLINE");
                }
            }

            @Override
            public void onFailure(Call<ProductsData> call, Throwable t) {
                networkStatus.setText("OFFLINE");
            }
        });
    }
    private Product getProductById(int id){
        Product product = null;
        for (int i = 0; i < productList.size(); i++){
            if(productList.get(i).getId() == id){
                product = productList.get(i);
            }
        }
        return product;
    }

    private void setSum(Position position){
        if ((position.getCount()!= null)&&(position.getPrice()!=null)&&(position.getDiscount()!=null)) {
            position.setSum((position.getPrice() - position.getDiscount()) * position.getCount());
        }
    }
}

File: src/main/java/controllers/cashbox/CashboxViewController.kt
package controllers.cashbox

import api.API
import api.APIMiddlewareImpl
import api.State
import api.StateListener
import javafx.collections.FXCollections
import javafx.event.ActionEvent
import javafx.event.EventHandler
import javafx.fxml.FXML
import javafx.fxml.FXMLLoader
import javafx.scene.Node
import javafx.scene.Parent
import javafx.scene.Scene
import javafx.scene.control.Label
import javafx.scene.control.TableColumn
import javafx.scene.control.TableView
import javafx.scene.control.cell.PropertyValueFactory
import javafx.stage.Modality
import javafx.stage.Stage
import kotlinx.coroutines.experimental.javafx.JavaFx
import kotlinx.coroutines.experimental.launch
import models.Operation
import models.Position
import models.Product
import start.Main
import java.io.IOException

/**
 * Created by andrey on 24.07.17.
 */
class CashboxViewController {
    private val mainApp: Main? = null

    internal var api: API = APIMiddlewareImpl

    private var operationOL = FXCollections.observableArrayList<Operation>()
    private var operationList: List<Operation>? = null
    private var productList: List<Product>? = null

    @FXML private lateinit var networkStatus: Label
    @FXML private lateinit var operationTable: TableView<Operation>
    @FXML private lateinit var typeColumn: TableColumn<Operation, String>
    @FXML private lateinit var dateColumn: TableColumn<Operation, String>
    @FXML private lateinit var userColumn: TableColumn<Operation, String>
    @FXML private lateinit var totalLable: Label

    @FXML private lateinit var userLable: Label
    @FXML private lateinit var typeLable: Label
    @FXML private lateinit var dateLable: Label

    @FXML private lateinit var positionTable: TableView<Position>
    @FXML private lateinit var nameColumn: TableColumn<Position, String>
    @FXML private lateinit var priceColumn: TableColumn<Position, Float>
    @FXML private lateinit var unitColumn: TableColumn<Position, String>
    @FXML private lateinit var countColumn: TableColumn<Position, String>
    @FXML private lateinit var discountColumn: TableColumn<Position, String>
    @FXML private lateinit var sumColumn: TableColumn<Position, String>


    @FXML
    private fun initialize() {

        nameColumn.setCellValueFactory(PropertyValueFactory("productName"))
        priceColumn.setCellValueFactory(PropertyValueFactory("priceFormat"))
        countColumn.setCellValueFactory(PropertyValueFactory("countString"))
        discountColumn.setCellValueFactory(PropertyValueFactory("discountString"))
        sumColumn.setCellValueFactory(PropertyValueFactory("sumFormat"))
        unitColumn.setCellValueFactory(PropertyValueFactory("unit"))
        //---------------------------------------------------------------------------------------
        typeColumn.setCellValueFactory(PropertyValueFactory("typeString"))
        dateColumn.setCellValueFactory(PropertyValueFactory("dateFormat"))
        userColumn.setCellValueFactory(PropertyValueFactory("user"))
        operationTable.selectionModel.selectedItemProperty().addListener { _, _, newValue -> showOperation(newValue) }
        loadOperationData()
        loadProducts()
        //---------------------------------------------------------------------------------------

        api.addStateListener(object : StateListener {
            override fun stateChanged(state: State) {
                networkStatus.text = state.toString()
            }
        })
    }

    private fun loadOperationData() = launch(JavaFx) {
        val result = api.operations().await()
        if (result.isSuccessful()) {
            operationList = result.result
            operationOL = FXCollections.observableArrayList(result.result)
            operationTable.setItems(operationOL)
        }
    }

    fun newOperation(actionEvent: ActionEvent) {
        try {
            val stage = Stage()
            val loader = FXMLLoader()
            loader.location = javaClass.getResource("/view/cashbox/NewCashboxView.fxml")
            val categoryAddFXML = loader.load<Parent>()
            stage.title = "Кассовая операция"
            stage.minHeight = 600.0
            stage.minWidth = 800.0
            stage.isResizable = false
            stage.scene = Scene(categoryAddFXML)
            stage.initModality(Modality.WINDOW_MODAL)
            stage.initOwner((actionEvent.source as Node).scene.window)
            val controller = loader.getController<NewCashboxController>()
            stage.onCloseRequest = EventHandler { }
            stage.showAndWait()

            if (controller.isOkClicked) {
                loadOperationData()
            }

        } catch (e: IOException) {
            e.printStackTrace()
        }

    }

    private fun showOperation(operation: Operation) {
        userLable.text = operation.user
        typeLable.text = operation.typeString
        dateLable.text = operation.dateFormat
        val positions = operation.positions
        var sum = 0.0
        positions.forEach { position ->
            getProductById(position.product)?.let {
                position.productName = it.name
                setSum(position)
                it.unit = it.unit
                sum += position.sum
            }
        }
        val positionObservableList = FXCollections.observableArrayList(operation.positions)
        positionTable.items = positionObservableList
        totalLable.text = String.format("%.2f", sum)
    }

    private fun loadProducts() = launch(JavaFx) {
        val result = api.productsData().await()
        if (result.isSuccessful()) {
            productList = result.result?.products
        }
    }

    private fun getProductById(id: Int) = productList?.firstOrNull { it.id == id }

    private fun setSum(position: Position) {
        with(position) {
            val count = count
            val price = price
            val discount = discount

            if (count != null && price != null && discount != null) {
                sum = (price - discount) * count
            }
        }
    }
}
